### PR TITLE
Compute non custom build and non transitive deps for doc

### PR DESCRIPTION
**Pull request by hi-rustin**
### What does this PR try to resolve?

close https://github.com/rust-lang/cargo/issues/10318 and close https://github.com/rust-lang/cargo/issues/9198
### How should we test and review this PR?

Compute non custom build and non transitive deps for doc. 
Add test for it.

### CODE REVIEW

**Comment by hi-rustin** [comment]
r? @ehuss 

The robot doesn't seem to be working.
Could you please take a look? Thanks! 💚 💙 💜 💛 ❤️

**Review by ehuss — COMMENTED** [review]
Thanks!

I think since the code is now identical between both, can all of this filtering be moved to `state.deps`?  Instead of passing in a callback, just inline the code where `filter` is called.

**Comment by hi-rustin** [comment]
> I think since the code is now identical between both, can all of this filtering be moved to `state.deps`? Instead of passing in a callback, just inline the code where `filter` is called.

Make sense. Addressed!

File: src/cargo/core/compiler/unit_dependencies.rs [hunk 1]
```diff
@@ -830,21 +807,44 @@ impl<'a, 'cfg> State<'a, 'cfg> {
             .unwrap_or_else(|_| panic!("expected {} to be downloaded", id))
     }
 
-    /// Returns a filtered set of dependencies for the given unit.
-    fn deps(
+    /// Returns non-custom and non-transitive dependencies.
+    fn non_custom_build_and_non_transitive_deps(
```
Comment (ehuss) [inline src/cargo/core/compiler/unit_dependencies.rs:811]: Can you keep the method name the same?  It is understood that this method is already filtering some deps.

File: src/cargo/core/compiler/unit_dependencies.rs [hunk 2]
```diff
@@ -830,21 +807,44 @@ impl<'a, 'cfg> State<'a, 'cfg> {
...
         &self,
         unit: &Unit,
         unit_for: UnitFor,
-        filter: &dyn Fn(&Dependency) -> bool,
     ) -> Vec<(PackageId, &HashSet<Dependency>)> {
+        let non_custom_build_and_non_transitive_filter = |dep: &Dependency| {
```
[the remark this answers is not in the document]
Comment (hi-rustin) [reply src/cargo/core/compiler/unit_dependencies.rs]: Addressed!

**Comment by ehuss** [comment]
Thanks!  There was once a reason why these two different code paths had different filters, but I cannot recall from the dusty recesses of my mind why that was. I've tried to think of any potential problems this might cause, but I'm unable to think of any scenario where it won't work.

@bors r+

**Pull request merged**